### PR TITLE
Implement fast (A* based) solver for goap

### PR DESCRIPTION
**Pull request by antoinealb**
Only requires minor changes of the API, which I did not do for the robot.

Basically we add a "distance" which tries to know how far we are from the objective. I am not sure if this is 100% optimal (probably not, depends on the distance function). However the Python implementation of this algorithm is 100x faster on the current GameGoal, so I think it might be worth it to test it.

### CODE REVIEW

**Comment by antoinealb** [comment]
This is not complete yet and should not be used . However it already gives answers and we could test its speed to see if it is worth finishing the work :)

File: lib/goap/goap.hpp [hunk 1]
```diff
@@ -22,19 +24,27 @@ class Action {
...
     virtual bool execute(State &state) = 0;
+
 };
 
 template<typename State>
 class Goal {
 public:
     /** Checks if the goal is reached for the given state. */
-    virtual bool is_reached(State state) = 0;
+    virtual bool is_reached(const State &state) const
```
[the remark this answers is not in the document]
Comment (antoinealb) [reply lib/goap/goap.hpp]: hmm maybe, i did not see a reason not to allow classes to do it.

File: lib/goap/goap_internals.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+template<typename State>
+void visited_states_array_to_list(VisitedState<State> *nodes, int len)
+{
+    int i;
+    for (i = 0; i < len-1; i++) {
+        nodes[i].next = &nodes[i+1];
+    }
+    nodes[len-1].next = nullptr;
```
[the remark this answers is not in the document]
Comment (antoinealb) [reply lib/goap/goap_internals.hpp]: don't think so, I think nullptr is pretty much self explaining here :)

File: lib/goap/goap.hpp [hunk 3]
```diff
@@ -51,58 +61,61 @@ class Planner {
     int plan(State state, Goal<State> &goal,
              Action<State> **path = nullptr, int path_len = 10)
```
[the remark this answers is not in the document]
Comment (antoinealb) [reply lib/goap/goap.hpp]: yeah, i plan to make the path an iterator (or iterator ish if its too hard).

File: lib/goap/goap.hpp [hunk 4]
```diff
@@ -51,58 +61,61 @@ class Planner {
...
-        if (path != nullptr) {
-            /* First get the path, then prepends our solution. */
-            State next_state = best_action->plan_effects(state);
-            plan(next_state, goal, &path[1], path_len - 1);
-            path[0] = best_action;
         }
 
-        return min_cost + 1;
+        // No path was found
+        return -1;
```
[the remark this answers is not in the document]
Comment (antoinealb) [reply lib/goap/goap.hpp]: return 0 means you are already at a satisfied goal

File: lib/goap/goap.hpp [hunk 5]
```diff
@@ -51,58 +61,61 @@ class Planner {
...
 
-        for (size_t i = 0; i < action_count; i++) {
+        while (open) {
+            auto current = priority_list_pop(open);
 
-            /* If the action cannot be used, then abort. */
-            if (actions[i]->can_run(state) == false) {
-                continue;
+            if (goal.is_reached(current->state)) {
+                // TODO provide something better for path, maybe iterator like ?
```
[the remark this answers is not in the document]
Comment (antoinealb) [reply lib/goap/goap.hpp]: for the resulting path.

**Comment by antoinealb** [comment]
I dont know why it fails, but the algorithm is not complete yet, so maybe there is that. Its just a rough draft, or WIP.

File: lib/goap/goap.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SyrianSpock) [reply lib/goap/goap.hpp]: mmh I see

**Comment by antoinealb** [comment]
Ok the failing unit test was the result of  building a bad distance metric.

File: lib/goap/goap.hpp [hunk 6]
```diff
@@ -97,19 +100,54 @@ class Planner {
...
+                    for (auto p = close; p; p = p->next) {
+                        if (p->state == neighbor->state) {
+                            should_insert = false;
+                            if (p->cost > neighbor->cost) {
+                                p->priority = neighbor->cost;
+                                p->priority = neighbor->priority;
+                                p->parent = neighbor->parent;
+                                p->action = neighbor->action;
+                            }
+                        }
```
Comment (SyrianSpock) [inline lib/goap/goap.hpp]: Same block as above repeated, maybe hide in a function.

File: lib/goap/tests/goap_internals.cpp [hunk 7]
```diff
@@ -75,3 +83,11 @@ TEST(InternalVisitedListState, CanPushListHead)
     POINTERS_EQUAL(&nodes[0], head->next);
 }
 
+TEST(InternalVisitedListState, CanPushEmptyList)
+{
+    VisitedState<MyState> new_elem, *head=nullptr;
```
Comment (SyrianSpock) [inline lib/goap/tests/goap_internals.cpp]: [nitpicking] Move declaration of head in new line

**Pull request merged**